### DISCUSSION TITLE
Several eval metrics and losses

### DISCUSSION THREAD

**Discussion by alibek-wand** [autogluon/autogluon#2583] [Q&A] [answered]
Hi!

Is is possible in TabularPredictor train each model, for example LGBM, on "log_loss", but train and tune weighted ensemble of models using another metric, let's say "accuracy"?

Thanks)

**Comment by Innixma** [comment, marked as answer]
Yes. Use `stopping_metric` as shown in the code example below. The `stopping_metric` of the weighted ensemble is equal to the `eval_metric` specified in `predictor = TabularPredictor(label=label, eval_metric='roc_auc')`.

```python3
from autogluon.tabular import TabularPredictor, TabularDataset


if __name__ == '__main__':
    path_prefix = 'https://autogluon.s3.amazonaws.com/datasets/AdultIncomeBinaryClassification/'
    path_train = path_prefix + 'train_data.csv'
    path_test = path_prefix + 'test_data.csv'

    label = 'class'
    train_data = TabularDataset(path_train)
    test_data = TabularDataset(path_test)

    hyperparameters = {
        'GBM': [
            {},
            {'ag_args_fit': {'stopping_metric': 'accuracy'}},
            {'ag_args_fit': {'stopping_metric': 'f1'}},
        ]
    }

    predictor = TabularPredictor(label=label, eval_metric='roc_auc')
    predictor.fit(train_data=train_data, hyperparameters=hyperparameters, verbosity=3)
    leaderboard = predictor.leaderboard(test_data)
```

Output:

```
                 model  score_test  score_val  pred_time_test  pred_time_val  fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0           LightGBM_3    0.930780   0.938036        0.031640       0.009637  1.213864                 0.031640                0.009637           1.213864            1       True          3
1           LightGBM_2    0.930780   0.938036        0.039179       0.009692  0.532410                 0.039179                0.009692           0.532410            1       True          2
2  WeightedEnsemble_L2    0.930570   0.938446        0.150463       0.023749  2.578784                 0.003718                0.000658           0.174269            2       True          4
3             LightGBM    0.930059   0.938331        0.107566       0.013399  1.872105                 0.107566                0.013399           1.872105            1       True          1
```

Example training log showing the different stopping metrics being used:

```
Model configs that will be trained (in order):
	LightGBM: 	{'ag_args': {'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 90}}
	LightGBM_2: 	{'ag_args_fit': {'stopping_metric': 'accuracy'}, 'ag_args': {'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 90}}
	LightGBM_3: 	{'ag_args_fit': {'stopping_metric': 'f1'}, 'ag_args': {'model_type': <class 'autogluon.tabular.models.lgb.lgb_model.LGBModel'>, 'priority': 90}}
Fitting 3 L1 models ...
Fitting model: LightGBM ...
	Fitting LightGBM with 'num_gpus': 0, 'num_cpus': 6
	Fitting 10000 rounds... Hyperparameters: {'learning_rate': 0.05}
[50]	valid_set's binary_logloss: 0.291482
[100]	valid_set's binary_logloss: 0.270547
[150]	valid_set's binary_logloss: 0.266672
[200]	valid_set's binary_logloss: 0.264774
[250]	valid_set's binary_logloss: 0.263957
[300]	valid_set's binary_logloss: 0.263742
[350]	valid_set's binary_logloss: 0.263563
[400]	valid_set's binary_logloss: 0.263068
[450]	valid_set's binary_logloss: 0.262766
[500]	valid_set's binary_logloss: 0.263105
[550]	valid_set's binary_logloss: 0.263881
[600]	valid_set's binary_logloss: 0.263983
Saving AutogluonModels/ag-20221220_173257/models/LightGBM/model.pkl
Saving AutogluonModels/ag-20221220_173257/utils/attr/LightGBM/y_pred_proba_val.pkl
	0.9383	 = Validation score   (roc_auc)
	1.87s	 = Training   runtime
	0.01s	 = Validation runtime
Saving AutogluonModels/ag-20221220_173257/models/trainer.pkl
Fitting model: LightGBM_2 ...
	Fitting LightGBM_2 with 'num_gpus': 0, 'num_cpus': 6
	Fitting 10000 rounds... Hyperparameters: {'learning_rate': 0.05}
[50]	valid_set's binary_error: 0.1272
[100]	valid_set's binary_error: 0.1204
[150]	valid_set's binary_error: 0.1208
[200]	valid_set's binary_error: 0.1192
[250]	valid_set's binary_error: 0.1196
[300]	valid_set's binary_error: 0.1184
[350]	valid_set's binary_error: 0.1192
Saving AutogluonModels/ag-20221220_173257/models/LightGBM_2/model.pkl
Saving AutogluonModels/ag-20221220_173257/utils/attr/LightGBM_2/y_pred_proba_val.pkl
	0.938	 = Validation score   (roc_auc)
	0.53s	 = Training   runtime
	0.01s	 = Validation runtime
Saving AutogluonModels/ag-20221220_173257/models/trainer.pkl
Fitting model: LightGBM_3 ...
	Fitting LightGBM_3 with 'num_gpus': 0, 'num_cpus': 6
	Fitting 10000 rounds... Hyperparameters: {'learning_rate': 0.05}
[50]	valid_set's binary_logloss: 0.291482	valid_set's f1: 0.706642
[100]	valid_set's binary_logloss: 0.270547	valid_set's f1: 0.730528
[150]	valid_set's binary_logloss: 0.266672	valid_set's f1: 0.73227
[200]	valid_set's binary_logloss: 0.264774	valid_set's f1: 0.737213
[250]	valid_set's binary_logloss: 0.263957	valid_set's f1: 0.736099
[300]	valid_set's binary_logloss: 0.263742	valid_set's f1: 0.738977
[350]	valid_set's binary_logloss: 0.263563	valid_set's f1: 0.736283
Saving AutogluonModels/ag-20221220_173257/models/LightGBM_3/model.pkl
Saving AutogluonModels/ag-20221220_173257/utils/attr/LightGBM_3/y_pred_proba_val.pkl
	0.938	 = Validation score   (roc_auc)
	1.21s	 = Training   runtime
	0.01s	 = Validation runtime
```

**Comment by alibek-wand** [comment]
Thanks a lot)
